modules/common/daemonset/daemonset.go

```diff
@@ -0,0 +1,134 @@
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openstack-k8s-operators/lib-common/modules/common/helper"
+	"github.com/openstack-k8s-operators/lib-common/modules/common/util"
+	appsv1 "k8s.io/api/apps/v1"
+	k8s_errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+)
+
+// NewDaemonSet returns an initialized DaemonSet
+func NewDaemonSet(
+	daemonset *appsv1.DaemonSet,
+	timeout int,
+) *DaemonSet {
+	return &DaemonSet{
+		daemonset: daemonset,
+		timeout:   time.Duration(timeout) * time.Second,
+	}
+}
+
+// CreateOrPatch - creates or patches a deployment, reconciles after Xs if object won't exist.
+func (d *DaemonSet) CreateOrPatch(
+	ctx context.Context,
+	h *helper.Helper,
+) (ctrl.Result, error) {
+	daemonset := &appsv1.DaemonSet{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      d.daemonset.Name,
+			Namespace: d.daemonset.Namespace,
+		},
+	}
+
+	op, err := controllerutil.CreateOrPatch(ctx, h.GetClient(), daemonset, func() error {
+		// Daemonset selector is immutable so we set this value only if
+		// a new object is going to be created
+		if daemonset.ObjectMeta.CreationTimestamp.IsZero() {
+			daemonset.Spec.Selector = d.daemonset.Spec.Selector
+		}
+		daemonset.Annotations = util.MergeStringMaps(daemonset.Annotations, d.daemonset.Annotations)
+		daemonset.Labels = util.MergeStringMaps(daemonset.Labels, d.daemonset.Labels)
+		daemonset.Spec.Template = d.daemonset.Spec.Template
+
+		err := controllerutil.SetControllerReference(h.GetBeforeObject(), daemonset, h.GetScheme())
+		if err != nil {
+			return err
+		}
+
+		return nil
+	})
+	if err != nil {
+		if k8s_errors.IsNotFound(err) {
+			h.GetLogger().Info(fmt.Sprintf("DaemonSet %s not found, reconcile in %s", daemonset.Name, d.timeout))
+			return ctrl.Result{RequeueAfter: d.timeout}, nil
+		}
+		return ctrl.Result{}, err
+	}
+	if op != controllerutil.OperationResultNone {
+		h.GetLogger().Info(fmt.Sprintf("DaemonSet %s - %s", daemonset.Name, op))
+	}
+
+	// update the daemonset object of the daemonset type
+	d.daemonset, err = GetDaemonSetWithName(ctx, h, daemonset.GetName(), daemonset.GetNamespace())
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// Delete - delete a daemonset.
+func (d *DaemonSet) Delete(
+	ctx context.Context,
+	h *helper.Helper,
+) error {
+	err := h.GetClient().Delete(ctx, d.daemonset)
+	if err != nil && !k8s_errors.IsNotFound(err) {
+		err = fmt.Errorf("error deleting daemonset %s: %v", d.daemonset.Name, err)
+		return err
+	}
+
+	return nil
+}
+
+// GetDaemonSet - get the daemonset object.
+func (d *DaemonSet) GetDaemonSet() appsv1.DaemonSet {
+	return *d.daemonset
+}
+
+// SetTimeout defines the duration used for requeueing while waiting for the daemonset
+// to finish.
+func (d *DaemonSet) SetTimeout(timeout time.Duration) {
```

Comment:
I would remove this and use the NewDaemonSet with a time.Duration parameter instead.